test/topology_custom/test_mv_admission_control.py

```diff
@@ -0,0 +1,132 @@
+#
+# Copyright (C) 2024-present ScyllaDB
+#
+# SPDX-License-Identifier: AGPL-3.0-or-later
+#
+from test.pylib.manager_client import ManagerClient
+
+import asyncio
+import pytest
+import time
+import logging
+
+from test.topology.conftest import skip_mode
+from test.pylib.util import wait_for_view
+from test.topology_experimental_raft.test_mv_tablets import pin_the_only_tablet, get_tablet_replicas
+
+from cassandra.cluster import ConsistencyLevel, EXEC_PROFILE_DEFAULT # type: ignore
+from cassandra.cqltypes import Int32Type # type: ignore
+from cassandra.policies import FallthroughRetryPolicy # type: ignore
+from cassandra.query import SimpleStatement, BoundStatement # type: ignore
+
+logger = logging.getLogger(__name__)
+
+# In the test, we create a table and perform a pair of writes to it. The
+# second write should fail due to admission control. We check that this
+# is indeed the error thrown.
+@pytest.mark.asyncio
+@skip_mode('release', "error injections aren't enabled in release mode")
+async def test_mv_admission_control_exception(manager: ManagerClient) -> None:
+    node_count = 2
+    config = {'error_injections_at_startup': ['view_update_limit', 'delay_before_remote_view_update', 'update_backlog_immediately'], 'enable_tablets': True}
+    servers = await manager.servers_add(node_count, config=config)
+    cql, hosts = await manager.get_ready_cql(servers)
+    await cql.run_async(f"CREATE KEYSPACE ks WITH replication = {{'class': 'NetworkTopologyStrategy', 'replication_factor': 1}}"
+                        "AND tablets = {'initial': 1}")
+    await cql.run_async(f"CREATE TABLE ks.tab (key int, c int, v text, PRIMARY KEY (key, c))")
+    await cql.run_async(f"CREATE MATERIALIZED VIEW ks.mv_cf_view AS SELECT * FROM ks.tab "
+                    "WHERE c IS NOT NULL and key IS NOT NULL PRIMARY KEY (c, key) ")
+    await wait_for_view(cql, 'mv_cf_view', node_count)
+
+    # Only remote updates hold on to memory, so make the update remote by pinning base and view tablets to different nodes.
+    await pin_the_only_tablet(manager, "ks", "tab", servers[0])
+    await pin_the_only_tablet(manager, "ks", "mv_cf_view", servers[1])
+
+    # Prepare the statement so that the write goes to the same shard both
+    # times (the first write will cause only the shard on which it was
+    # performed to have the updated view update backlog).
```

Comment:
This "goes to the same shard" might only work with a specific choice of CQL driver (Scylla's one). Do you really need this assumption? What happens if it isn't true (e.g., I run the test using Cassandra's driver) - will the test fail, or just fail to reproduce the bug, or what?

Reply:
Unfortunately, we need this assumption because the information about the increased backlog will only be received on the shard that sends (and receives response for) the first write. With a different (non-shard aware) driver, the increased view backlog will not be noticed so we won't get the admission control error which we expect. If in the future we'll want to try running this with Cassandra's driver, we'll probably need to either skip this test, propagate the received backlogs to other shards, or find another way to send requests to a specific Scylla shard